src/Elm/Parser/Base.elm

```diff
@@ -5,15 +5,32 @@ import Elm.Parser.Tokens as Tokens
 import Elm.Syntax.ModuleName exposing (ModuleName)
 import Elm.Syntax.Node exposing (Node)
 import Parser as Core exposing ((|.), (|=))
-import Parser.Extra
 
 
 moduleName : Core.Parser (Node ModuleName)
 moduleName =
-    Parser.Extra.sepBy1 "." Tokens.typeName
+    Core.succeed listCons
+        |= Tokens.typeName
+        |= moduleNameOrEmpty
         |> Node.parserCore
 
 
+listCons : a -> List a -> List a
+listCons head =
+    \tail -> head :: tail
+
+
+moduleNameOrEmpty : Core.Parser ModuleName
+moduleNameOrEmpty =
+    Core.oneOf
+        [ Core.succeed listCons
+            |. Tokens.dot
+            |= Tokens.typeName
+            |= Core.lazy (\() -> moduleNameOrEmpty)
+        , Core.lazy (\() -> Core.succeed [])
```

Comment:
Using Core.succeed [] is probably fine, seeing as it will only ever be evaluated once.